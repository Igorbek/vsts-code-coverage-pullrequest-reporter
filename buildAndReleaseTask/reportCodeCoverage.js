"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const BuildInterfaces_1 = require("azure-devops-node-api/interfaces/BuildInterfaces");
const GitInterfaces_1 = require("azure-devops-node-api/interfaces/GitInterfaces");
//import { CodeCoverageStatistics } from "vso-node-api/interfaces/TestInterfaces";
const messageFormatter_1 = require("./messageFormatter");
function reportCodeCoverage(vssConnection, buildId, projectId) {
    return __awaiter(this, void 0, void 0, function* () {
        const testApi = yield vssConnection.getTestApi();
        const codeCoverage = yield testApi.getCodeCoverageSummary(projectId, buildId);
        if (!codeCoverage || !codeCoverage.coverageData || !codeCoverage.coverageData.length) {
            console.log(`No code coverage data for build #${buildId}.`);
        }
        const buildApi = yield vssConnection.getBuildApi();
        const build = yield buildApi.getBuild(projectId, buildId);
        if (build.reason !== BuildInterfaces_1.BuildReason.PullRequest) {
            console.log(`The reason of build #${buildId} is not pull request.`);
            return;
        }
        const pullRequestId = getPullRequestIdFromBranchName(build.sourceBranch);
        if (!pullRequestId) {
            console.log(`Unable to determine pull request id from branch name '${build.sourceBranch}'.`);
            return;
        }
        const codeApi = yield vssConnection.getGitApi();
        const pullRequest = yield codeApi.getPullRequestById(pullRequestId);
        if (!pullRequest) {
            console.log(`Pull request #${pullRequestId} does not exist.`);
            return;
        }
        const lastMergeTargetCommitId = pullRequest.lastMergeTargetCommit.commitId;
        const builds = yield buildApi.getBuilds(projectId, [build.definition.id], undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, 10, undefined, undefined, undefined, undefined, pullRequest.targetRefName);
        if (!builds || builds.length < 1) {
            console.log(`No builds found for target branch '${pullRequest.targetRefName}'.`);
            return;
        }
        const lastMergeTargetBuild = builds.find(b => b.sourceVersion === lastMergeTargetCommitId);
        if (!lastMergeTargetBuild) {
            console.log(`Build of merge target commit does not exist.`);
            return;
        }
        const targetCodeCoverage = yield testApi.getCodeCoverageSummary(projectId, lastMergeTargetBuild.id);
        if (!targetCodeCoverage || !targetCodeCoverage.coverageData || !targetCodeCoverage.coverageData.length) {
            console.log(`No code coverage for target branch '${pullRequest.targetRefName}'.`);
            return;
        }
        const message = `### Code Coverage Report\n${messageFormatter_1.formatMarkdownReport({ label: 'Target', stat: targetCodeCoverage.coverageData[0] }, { label: 'PR', stat: codeCoverage.coverageData[0] })}`;
        const pullRequestThreads = yield codeApi.getThreads(pullRequest.repository.id, pullRequestId);
        let pullRequestThread = pullRequestThreads.find(prThread => prThread.properties && prThread.properties.codeCoverageReport);
        if (pullRequestThread && pullRequestThread.comments.length) {
            const comment = pullRequestThread.comments[0];
            yield codeApi.updateComment({
                commentType: GitInterfaces_1.CommentType.Text,
                content: message
            }, pullRequest.repository.id, pullRequestId, pullRequestThread.id, comment.id);
        }
        else {
            yield codeApi.createThread({
                comments: [{
                        commentType: GitInterfaces_1.CommentType.Text,
                        content: message
                    }],
                properties: { codeCoverageReport: true },
                status: GitInterfaces_1.CommentThreadStatus.Active,
            }, pullRequest.repository.id, pullRequestId);
        }
    });
}
exports.default = reportCodeCoverage;
function getPullRequestIdFromBranchName(branchName) {
    const match = branchName.match(/refs\/pull\/(\d+)\/merge/);
    return match && parseInt(match[1]) || undefined;
}
