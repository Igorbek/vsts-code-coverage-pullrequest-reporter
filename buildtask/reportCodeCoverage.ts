import { WebApi } from "vso-node-api/WebApi";
import { BuildReason } from "vso-node-api/interfaces/BuildInterfaces";

export default async function reportCodeCoverage(
  vssConnection: WebApi,
  buildId: number,
  projectId: string
) {
  const testApi = vssConnection.getTestApi();
  const codeCoverage = await testApi.getCodeCoverageSummary(projectId, buildId);

  codeCoverage.coverageData.forEach(d => console.log(d));

  const buildApi = vssConnection.getBuildApi();
  const build = await buildApi.getBuild(buildId);

  if (build.reason !== BuildReason.PullRequest) {
      return;
  }

  const pullRequestId = getPullRequestIdFromBranchName(build.sourceBranch);
  if (!pullRequestId) {
      // message
      return;
  }

  const codeApi = vssConnection.getGitApi();
  const pullRequest = await codeApi.getPullRequestById(pullRequestId);

  const builds = await buildApi.getBuilds(
    projectId,
    [build.definition.id],
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    1,
    undefined,
    undefined,
    undefined,
    undefined,
    pullRequest.targetRefName
  );

  if (!builds || builds.length < 1) {
      return;
  }

  const lastTargetBuildId = builds[0].id;

  const targetCodeCoverage = await testApi.getCodeCoverageSummary(projectId, lastTargetBuildId);

  targetCodeCoverage.coverageData.forEach(d => console.log(d));
}

function getPullRequestIdFromBranchName(branchName: string): number | undefined {
    const match = branchName.match(/refs\/pull\/(\d+)\/merge/);
    return match && parseInt(match[1]) || undefined;
}