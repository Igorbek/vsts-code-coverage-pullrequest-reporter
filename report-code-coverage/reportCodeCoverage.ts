import { WebApi } from "vso-node-api/WebApi";
import { BuildReason } from "vso-node-api/interfaces/BuildInterfaces";
import { CommentThreadStatus, CommentType } from "vso-node-api/interfaces/GitInterfaces";
import { CodeCoverageStatistics } from "vso-node-api/interfaces/TestInterfaces";
import { formatMarkdownReport } from './messageFormatter';

interface CodeCoverageStat {
    total: number;
    covered: number;
}

interface CodeCoverageComparison {
    target: CodeCoverageStat;
    source: CodeCoverageStat;
}

interface CodeCoverageComparisonEntries {
    [label: string]: CodeCoverageComparison;
}

export default async function reportCodeCoverage(
    vssConnection: WebApi,
    buildId: number,
    projectId: string
) {
    const testApi = vssConnection.getTestApi();
    const codeCoverage = await testApi.getCodeCoverageSummary(projectId, buildId);

    if (!codeCoverage || !codeCoverage.coverageData || !codeCoverage.coverageData.length) {
        console.log(`No code coverage data for build #${buildId}.`)
    }

    const buildApi = vssConnection.getBuildApi();
    const build = await buildApi.getBuild(buildId);

    if (build.reason !== BuildReason.PullRequest) {
        console.log(`The reason of build #${buildId} is not pull request.`);
        return;
    }

    const pullRequestId = getPullRequestIdFromBranchName(build.sourceBranch);
    if (!pullRequestId) {
        console.log(`Unable to determine pull request id from branch name '${build.sourceBranch}'.`);
        return;
    }

    const codeApi = vssConnection.getGitApi();
    const pullRequest = await codeApi.getPullRequestById(pullRequestId);

    if (!pullRequest) {
        console.log(`Pull request #${pullRequestId} does not exist.`);
        return;
    }

    const lastMergeTargetCommitId = pullRequest.lastMergeTargetCommit.commitId;

    const builds = await buildApi.getBuilds(
        projectId,
        [build.definition.id],
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        10,
        undefined,
        undefined,
        undefined,
        undefined,
        pullRequest.targetRefName
    );

    if (!builds || builds.length < 1) {
        console.log(`No builds found for target branch '${pullRequest.targetRefName}'.`);
        return;
    }

    const lastMergeTargetBuild = builds.find(b => b.sourceVersion === lastMergeTargetCommitId);
    if (!lastMergeTargetBuild) {
        console.log(`Build of merge target commit does not exist.`);
        return;
    }

    const targetCodeCoverage = await testApi.getCodeCoverageSummary(projectId, lastMergeTargetBuild.id);

    if (!targetCodeCoverage || !targetCodeCoverage.coverageData || !targetCodeCoverage.coverageData.length) {
        console.log(`No code coverage for target branch '${pullRequest.targetRefName}'.`);
        return;
    }

    const message = `### Code Coverage Report\n${formatMarkdownReport(
        { label: 'Target', stat: targetCodeCoverage.coverageData[0] },
        { label: 'PR', stat: codeCoverage.coverageData[0] })}`;

    const pullRequestThreads = await codeApi.getThreads(pullRequest.repository.id, pullRequestId);
    let pullRequestThread = pullRequestThreads.find(prThread => prThread.properties && prThread.properties.codeCoverageReport);
    if (pullRequestThread && pullRequestThread.comments.length) {
        const comment = pullRequestThread.comments[0];
        console.log(`codeApi.updateComment({},
            pullRequest.repository.id=${pullRequest.repository.id},
            pullRequestId=${pullRequestId},
            pullRequestThread.id=${pullRequestThread.id},
            comment.id=${comment.id})`);
        await codeApi.updateComment({
                commentType: CommentType.Text,
                content: message
            } as any,
            pullRequest.repository.id,
            pullRequestId,
            pullRequestThread.id,
            comment.id);
    } else {
        await codeApi.createThread({
            comments: [{
                commentType: CommentType.Text,
                content: message
            }],
            properties: { codeCoverageReport: true },
            status: CommentThreadStatus.Active,
        } as any, pullRequest.repository.id, pullRequestId);
    }
}

function getPullRequestIdFromBranchName(branchName: string): number | undefined {
    const match = branchName.match(/refs\/pull\/(\d+)\/merge/);
    return match && parseInt(match[1]) || undefined;
}
