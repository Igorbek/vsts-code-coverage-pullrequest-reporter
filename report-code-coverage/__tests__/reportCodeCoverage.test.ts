import { WebApi, getPersonalAccessTokenHandler } from 'vso-node-api/WebApi';
import { getPullRequestIdFromBranchName } from '../reportCodeCoverage';

describe('getPullRequestIdFromBranchName', () => {
    it('extracts PR numbers', () => {
        const inputs = [
            "/refs/pull/123/merge",
            "/refs/pull/0345/merge",
            "/refs/pull/abc/merge",
            "/refs/pull/merge",
            "/refs/pull/123",
            "master"
        ];

        const data = inputs.map(input => ({ input, output: getPullRequestIdFromBranchName(input) });

        expect(data).toMatchSnapshot();
    });
});
