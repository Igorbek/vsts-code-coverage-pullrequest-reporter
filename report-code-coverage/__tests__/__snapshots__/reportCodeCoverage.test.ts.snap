// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPullRequestIdFromBranchName extracts PR numbers 1`] = `
Array [
  Object {
    "input": "/refs/pull/123/merge",
    "output": 123,
  },
  Object {
    "input": "/refs/pull/0345/merge",
    "output": 345,
  },
  Object {
    "input": "/refs/pull/abc/merge",
    "output": undefined,
  },
  Object {
    "input": "/refs/pull/merge",
    "output": undefined,
  },
  Object {
    "input": "/refs/pull/123",
    "output": undefined,
  },
  Object {
    "input": "master",
    "output": undefined,
  },
]
`;
